# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [3]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv',skiprows=4)

# TODO: remove the 'Unnamed:62' column from each data set
df_rural = df_rural.drop('Unnamed: 62', axis=1)
df_electricity = df_electricity.drop('Unnamed: 62', axis=1)
# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df = pd.concat([df_rural, df_electricity])
df.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965   ...    \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300   ...     
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250   ...     
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496   ...     
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770   ...     
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555   ...     

     2008    2009    2010    2011    2012    2013    2014    2015    2016  \
0  56.217  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696   
1  76.054  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868   
2  61.491  60.701  59.903  59.100  58.301  57.510  56.726  55.950  55.181   
3  50.009  48.924  47.837  46.753  45.670  44.617  43.591  42.593  41.624   
4  11.133  11.648  12.183  12.740  13.292  13.835  14.367  14.885  15.388   

     2017  
0  58.893  
1  72.430  
2  54.422  
3  40.684  
4  15.873  

[5 rows x 62 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [7]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_melt = pd.melt(df_rural, id_vars=["Country Name", "Country Code", "Indicator Name","Indicator Code"],\
                        var_name='Year', value_name="Rural_Value")

df_electricity_melt = pd.melt(df_electricity, id_vars = ["Country Name", "Country Code", "Indicator Name","Indicator Code"],\
                   var_name = "Year", value_name = "Electricity_Value")
# TODO: drop any columns from the data frames that aren't needed
df_rural_melt.drop(['Indicator Name', "Indicator Code"], axis = 1, inplace=True)
df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis = 1, inplace=True)
# TODO: merge the data frames together based on their common columns
df_me = pd.merge(df_rural_melt, df_electricity_melt, how='outer', on = ["Country Name", "Country Code", "Year"])
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

df_combined = df_me.sort_values(by = ["Country Name", "Year"])
df_combined.head()

Country Name Country Code  Year  Rural_Value  Electricity_Value
1     Afghanistan          AFG  1960       91.779                NaN
265   Afghanistan          AFG  1961       91.492                NaN
529   Afghanistan          AFG  1962       91.195                NaN
793   Afghanistan          AFG  1963       90.890                NaN
1057  Afghanistan          AFG  1964       90.574                NaN

In [11]:
df_combined[df_combined['Country Name'] == "Kenya"]

Country Name Country Code  Year  Rural_Value  Electricity_Value
119          Kenya          KEN  1960       92.638                NaN
383          Kenya          KEN  1961       92.435                NaN
647          Kenya          KEN  1962       92.226                NaN
911          Kenya          KEN  1963       91.962                NaN
1175         Kenya          KEN  1964       91.682                NaN
1439         Kenya          KEN  1965       91.395                NaN
1703         Kenya          KEN  1966       91.098                NaN
1967         Kenya          KEN  1967       90.791                NaN
2231         Kenya          KEN  1968       90.475                NaN
2495         Kenya          KEN  1969       90.150                NaN
2759         Kenya          KEN  1970       89.705                NaN
3023         Kenya          KEN  1971       89.222                NaN
3287         Kenya          KEN  1972       88.718                NaN
3551         Kenya          KEN  1973       88.195                NaN
3815         Kenya          KEN  1974       87.651                NaN
4079         Kenya          KEN  1975       87.086                NaN
4343         Kenya          KEN  1976       86.497                NaN
4607         Kenya          KEN  1977       85.888                NaN
4871         Kenya          KEN  1978       85.255                NaN
5135         Kenya          KEN  1979       84.599                NaN
5399         Kenya          KEN  1980       84.417                NaN
5663         Kenya          KEN  1981       84.319                NaN
5927         Kenya          KEN  1982       84.220                NaN
6191         Kenya          KEN  1983       84.121                NaN
6455         Kenya          KEN  1984       84.021                NaN
6719         Kenya          KEN  1985       83.921                NaN
6983         Kenya          KEN  1986       83.820                NaN
7247         Kenya          KEN  1987       83.719                NaN
7511         Kenya          KEN  1988       83.617                NaN
7775         Kenya          KEN  1989       83.515                NaN
8039         Kenya          KEN  1990       83.252           3.318390
8303         Kenya          KEN  1991       82.957           4.614890
8567         Kenya          KEN  1992       82.658           5.910881
8831         Kenya          KEN  1993       82.355          10.900000
9095         Kenya          KEN  1994       82.048           8.490625
9359         Kenya          KEN  1995       81.737           9.768262
9623         Kenya          KEN  1996       81.421          11.033661
9887         Kenya          KEN  1997       81.102          12.283765
10151        Kenya          KEN  1998       80.778          14.500000
10415        Kenya          KEN  1999       80.450          14.725850
10679        Kenya          KEN  2000       80.108          15.913233
10943        Kenya          KEN  2001       79.761          17.082207
11207        Kenya          KEN  2002       79.409          18.238832
11471        Kenya          KEN  2003       79.052          16.000000
11735        Kenya          KEN  2004       78.690          20.539297
11999        Kenya          KEN  2005       78.325          21.695263
12263        Kenya          KEN  2006       77.955          22.863132
12527        Kenya          KEN  2007       77.580          24.048971
12791        Kenya          KEN  2008       77.200          25.257330
13055        Kenya          KEN  2009       76.817          23.000000
13319        Kenya          KEN  2010       76.429          19.200000
13583        Kenya          KEN  2011       76.033          28.996553
13847        Kenya          KEN  2012       75.630          30.271000
14111        Kenya          KEN  2013       75.220          31.554457
14375        Kenya          KEN  2014       74.803          36.000000
14639        Kenya          KEN  2015       74.378          41.600000
14903        Ken

In [10]:
df_electricity_melt.head()

Country Name Country Code  Year  Electricity_Value
0        Aruba          ABW  1960                NaN
1  Afghanistan          AFG  1960                NaN
2       Angola          AGO  1960                NaN
3      Albania          ALB  1960                NaN
4      Andorra          AND  1960                NaN